File: Leetcode/1358. Number of Substrings Containing All Three Characters/solution.py
def numberOfSubstrings(s: str) -> int:
  a = 0
  b = 0
  c = 0

  right = 0

  count = 0

  for index, char in enumerate(s):
    if char == "a":
      a += 1
    elif char == "b":
      b += 1
    else:
      c += 1

    while a > 0 and b > 0 and c > 0:
      count += len(s) - index

      if s[right] == "a":
        a -= 1
      elif s[right] == "b":
        b -= 1
      else:
        c -= 1

      right += 1

  return count

print(numberOfSubstrings("abcabc"))

File: Leetcode/1358. Number of Substrings Containing All Three Characters/solution.rb
def number_of_substrings(s)
  a = 0
  b = 0
  c = 0

  right = 0

  count = 0

  s.each_char.with_index do |char, index|
    if char == "a"
      a += 1
    elsif char == "b"
      b += 1
    else
      c += 1
    end

    while a > 0 && b > 0 && c > 0
      count += s.length - index
      a -= 1 if s[right] == "a"
      b -= 1 if s[right] == "b"
      c -= 1 if s[right] == "c"
      right += 1
    end
  end

  count
end

puts number_of_substrings("abcabc") // 10
